akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ClusterActorSetImpl.scala

```diff
@@ -0,0 +1,178 @@
+          val startedChild = context.spawn(factory(entityContext.entityId), "entity")
+          context.watch(startedChild)
+          child = Some(startedChild)
+        }
+
+        context.log.debug(s"Starting ClusterActorSet actor ${entityContext.entityId}")
+        startChild()
+
+        Behaviors
+          .receiveMessage[Command] {
+            case Stop =>
+              child match {
+                case Some(ref) =>
+                  ref ! childStopMessage
+                  waitingForChildToStop()
+                case None =>
+                  Behaviors.stopped
+              }
+
+            case Ping =>
+              Behaviors.same
+          }
+          .receiveSignal {
+            case (_, Terminated(_)) =>
+              // if the child stopped manually, but we are not being re-located to another node,
+              // we can fast-restart it instead of waiting for a ping
+              context.log.debug(s"ClusterActorSet actor ${entityContext.entityId} stopped, starting again")
+              startChild()
+              Behaviors.same
+          }
+      }
+
+    private def waitingForChildToStop(): Behavior[Command] = Behaviors.setup { context =>
+      // FIXME timeout and force termination if child doesn't stop?
+      Behaviors.receiveSignal {
+        case (_, Terminated(_)) =>
+          // if the child stopped manually, but we are not being re-located to another node,
+          // we can fast-restart it instead of waiting for a ping
+          context.log.debug(s"ClusterActorSet actor stopped")
+          Behaviors.stopped
+      }
+    }
+  }
+
+  object KeepAlivePinger {
+    sealed trait Event
+    case object Tick extends Event
+
+    def apply(
+        settings: ClusterActorSetSettings,
+        identities: Set[EntityId],
+        shardingRef: ActorRef[ShardingEnvelope[EntityParent.Command]]): Behavior[Event] =
+      Behaviors.setup { context =>
+        Behaviors.withTimers { timers =>
+          context.log.debug(
+            s"Starting ClusterActorSet KeepAlivePinger with ping interval ${settings.keepAliveInterval}")
+          // FIXME should we delay initial tick with random fraction of keepAliveInterval to avoid thundering herd?
+          timers.startTimerWithFixedDelay(Tick, settings.keepAliveInterval)
+          Behaviors.receiveMessage {
+            case Tick =>
+              identities.foreach(id => shardingRef ! ShardingEnvelope(id, EntityParent.Ping))
+              Behaviors.same
+          }
+        }
+      }
+  }
+
+}
+@InternalApi
+private[akka] class ClusterActorSetImpl(system: ActorSystem[_]) extends ClusterActorSet {
+
+  private val typeKeyCounter = new AtomicInteger(0)
+
+  override def init[T](numberOfEntities: Int, behaviorFactory: EntityId => Behavior[T], stopMessage: T): Unit =
+    init(ClusterActorSetSettings(system), numberOfEntities, behaviorFactory, stopMessage)
+
+  override def init[T](
+      settings: ClusterActorSetSettings,
+      numberOfEntities: Int,
```

Comment:
this is nice because number actors can be changed, in a rolling update

but we also need the `Set[String]` because the entityId will probably carry some business value needed by the actor, such as the tag slice identifier for the read side processors